controllers/componentdetectionquery_controller_conditions.go

```diff
@@ -61,9 +69,65 @@ func (r *ComponentDetectionQueryReconciler) SetCompleteConditionAndUpdateCR(ctx
 			Message: fmt.Sprintf("ComponentDetectionQuery failed: %v", completeError),
 		})
 	}
+	updateComponentName(ctx, componentDetectionQuery, r.Client)
 
 	err := r.Client.Status().Update(ctx, componentDetectionQuery)
 	if err != nil {
 		log.Error(err, "Unable to update ComponentDetectionQuery")
 	}
 }
+
+func updateComponentName(ctx context.Context, componentDetectionQuery *appstudiov1alpha1.ComponentDetectionQuery, client client.Client) {
+	for key, value := range componentDetectionQuery.Status.ComponentDetected {
+		repoUrl := value.ComponentStub.Source.GitSource.URL
+		lastElement := repoUrl[strings.LastIndex(repoUrl, "/")+1:]
+		repoName := strings.Split(lastElement, ".git")[0]
+		componentName := repoName
+		context := value.ComponentStub.Source.GitSource.Context
+		if len(componentDetectionQuery.Status.ComponentDetected) > 1 && context != "" && context != "./" {
+			componentName = fmt.Sprintf("%s-%s", context, repoName)
+		}
+		componentName = sanitizeComponentName(ctx, componentName, client, componentDetectionQuery.Namespace)
+		value.ComponentStub.ComponentName = componentName
+
+		componentDetectionQuery.Status.ComponentDetected[key] = value
+	}
+}
+
+// sanitizeComponentName sanitizes component name with the following requirements:
+// - Contain at most 63 characters
+// - Contain only lowercase alphanumeric characters or ‘-’
+// - Start with an alphanumeric character
+// - End with an alphanumeric character
+// - Must not contain all numeric values
+func sanitizeComponentName(ctx context.Context, name string, client client.Client, namespace string) string {
+	exclusive := regexp.MustCompile(`[^a-zA-Z0-9/-]`)
+	// filter out invalid characters
+	name = exclusive.ReplaceAllString(name, "")
+	_, err := strconv.ParseFloat(name, 64)
+	if err != nil {
+		// convert all Uppercase chars to lowercase
+		name = strings.ToLower(name)
+	} else {
+		// contains only numeric values, prefix a character
+		name = fmt.Sprintf("comp-%s", name)
+	}
+	if len(name) > 58 {
+		name = name[0:58]
+	}
+
+	// get hc
+	hc := &appstudiov1alpha1.Component{}
+	namespacedName := types.NamespacedName{
+		Name:      name,
+		Namespace: namespace,
+	}
+	newErr := client.Get(ctx, namespacedName, hc)
```

Comment:
I'm not crazy about us doing a Kubernetes API request each time we need to sanitize the name, which will be for the N components returned by a CDQ. There's already some concerns around the response time for CDQs, and this will likely slow it down further.

Could we just always add on the 4 random characters, and skip the calls to the Kubernetes API?

Reply:
I raised this question during our meeting, but last time Elson still wants to check for conflict instead of  always adding random character. 
I do agree the current CDQ response time is already very long when using HAC UI, and it's reasonable to avoid unnecessary slowing down. but the actual response time when running HAS through command line is not that bad.

We can discuss further when Elson comes back.

Reply:
While it may be fine via the command line, most of the users will be interacting with CDQs via HAC, so that's where the latency really matters IMO.

Let's discuss when Elson's back.